infra/growth/src/resources/rules/partnerReferralRule.js

```diff
@@ -0,0 +1,194 @@
+const fetch = require('cross-fetch')
+const { BaseRule } = require('./baseRule')
+const { Reward } = require('./reward')
+const logger = require('../../logger')
+const { tokenToNaturalUnits } = require('../../util/token')
+
+const CONF_CACHE_DURATION = 5 * 60 * 1000 // 5 minutes
+const PARTNER_CONF_URL =
+  process.env.PARTNER_CONF_URL ||
+  'https://originprotocol.com/static/partnerconf'
+const PARTNER_REWARDS = {}
+
+/**
+ * A rule that checks on a PartnerReferral event type.
+ */
+class PartnerReferralEvent extends BaseRule {
+  constructor(crules, levelId, config) {
+    super(crules, levelId, config)
+
+    if (this.config.eventType) {
+      this.addEventType(this.config.eventType)
+    }
+
+    this.validCodes = []
+    this.lastConfLoad = null
+  }
+
+  /**
+   * Return whether or not the configuration should be reloaded from source
+   */
+  async _reloadConf() {
+    if (
+      this.validCodes.length < 1 ||
+      (this.lastConfLoad &&
+        +new Date() - CONF_CACHE_DURATION > Number(this.lastConfLoad))
+    ) {
+      await this._getConfig()
+    }
+  }
+
+  /**
+   * Returns number of rewards user qualifies for, taking into account the rule's limit.
+   * @param {string} ethAddress - User's account.
+   * @param {Array<models.GrowthEvent>} events
+   * @returns {number}
+   * @private
+   */
+  async _numRewards(ethAddress, events) {
+    await this._reloadConf()
+    const tally = this._tallyEvents(
+      ethAddress,
+      this.eventTypes,
+      events,
+      customId => this.validCodes.includes(customId)
+    )
+    return tally && tally > 0 ? 1 : 0
+  }
+
+  /**
+   * Returns true if the rule passes, false otherwise.
+   * @param {string} ethAddress - User's account.
+   * @param {Array<models.GrowthEvent>} events
+   * @returns {boolean}
+   */
+  async _evaluate(ethAddress, events) {
+    await this._reloadConf()
+    const tally = this._tallyEvents(
+      ethAddress,
+      this.eventTypes,
+      events,
+      customId => this.validCodes.includes(customId)
+    )
+    return Object.keys(tally).length > 0 && Object.values(tally)[0] > 0
+  }
+
+  /**
+   * Retrieve all campaign configuration
+   */
+  async _getConfig() {
+    const url = `${PARTNER_CONF_URL}/campaigns.json`
+    const res = await fetch(url)
+    if (res.status !== 200) {
+      throw new Error('Failed to fetch campaigns JSON')
```

Comment:
wonder if we should log the error but not crash?
This way if for some reason originprotocol.com is down, it won't impact rewards (it'll use whatever last version of the config was fetched).

Reply:
Do you think anyone will see the logs?  If it can't load the config, it won't work.

Reply:
Made this change though...